src/relax/op/tensor/view.cc

```diff
@@ -0,0 +1,359 @@
+    // dtype is changing.  Any non-zero offset will cause the view to
+    // overrun the bounds of the original array.
+    if (ctx->GetAnalyzer()->CanProve(view_relative_byte_offset.value() > 0)) {
+      LOG(FATAL) << "ValueError: "
+                 << "Views into an array must not exceed the bounds of the array being viewed.  "
+                 << "However, expression " << call << " attempted to offset the view by "
+                 << view_relative_byte_offset << " bytes, "
+                 << "without reducing either the number of elements in the view "
+                 << "or the size of each element.";
+    }
+  }
+
+  if (output_shape.defined()) {
+    return TensorStructInfo(ShapeExpr(output_shape.value()), output_dtype, data_sinfo->vdevice);
+  } else {
+    return TensorStructInfo(output_dtype, output_ndim, data_sinfo->vdevice);
+  }
+}
+
+TVM_REGISTER_GLOBAL("tvm.relax.struct_info.infer_view_sinfo").set_body_typed(InferStructInfoView);
+
+Expr LegalizeView(const BlockBuilder& bb, const Call& call) {
+  Expr data = call->args[0];
+  Expr shape = call->args[1];
+  Expr dtype = call->args[2];
+  Expr relative_byte_offset = call->args[3];
+
+  if (HasVoidStructInfo(shape) && HasVoidStructInfo(dtype) &&
+      HasVoidStructInfo(relative_byte_offset)) {
+    // Special-case, no change is required by the view.
+    return data;
+  }
+
+  // Prior to legalization, it is useful to use void-type argument to
+  // specify "no change".  This allows for better shape inference when
+  // a pass updates the input `data` tensor.  However, when we
+  // legalize the `R.view`, we must provide an explicit parameters.
+
+  if (HasVoidStructInfo(shape)) {
+    auto data_shape = data->struct_info_.as<TensorStructInfo>().value()->GetShape();
+    CHECK(data_shape.defined())
+        << "Legalization of " << call->op
+        << " requires that either the output shape be explicitly specified, "
+        << "or the input shape is known.  "
+        << "However, in expression " << call << ", no output shape is specified, "
+        << "and the input " << data << " of type " << data->struct_info_ << " has unknown shape.";
+    shape = ShapeExpr(data_shape.value());
+  }
+
+  if (HasVoidStructInfo(dtype)) {
+    auto data_dtype = data->struct_info_.as<TensorStructInfo>().value()->dtype;
+    CHECK(!data_dtype.is_void())
+        << "Legalization of " << call->op
+        << " requires that either the output dtype be explicitly specified, "
+        << "or the input dtype is known.  "
+        << "However, in expression " << call << ", no output dtype is specified, "
+        << "and the input " << data << " of type " << data->struct_info_ << " has unknown dtype.";
+    dtype = relax::DataTypeImm(data_dtype);
+  }
+
+  if (HasVoidStructInfo(relative_byte_offset)) {
+    relative_byte_offset = relax::PrimValue::Int64(0);
+  }
+
+  StructInfoDeriveFunc infer_sinfo_env_func;
+  infer_sinfo_env_func = EnvFunc::Get("tvm.relax.struct_info.infer_view_sinfo");
```

Comment:
Why does this need to be a packed func? Can't we use the C++ function directly?

Reply:
If we want to define the StructInfo for the generated `Call` node, we could call the C++ function directly.  Ideally, though, if the arguments change due to some downstream transform, the shape inference should be repeated with the new argument.  For that

* Using an operator's `FInferStructInfo` function.  This only applies to `tvm.ir.Op` instances, and we've just removed that as part of `LegalizeOps`.
* Using the `params` and `ret` fields from the `FuncStructInfo`.  This works for static cases, and most dynamic shapes, but doesn't support inferring the output shape based on a `ShapeExpr` argument.
* Using `FuncStructInfo::OpaqueFunc` with a derivation func.  This is the most general method, and essentially lets you pack a `FInferStructInfo` into a `FuncStructInfo`.

The third option doesn't come up very often, and I could only find one previous location where this functionality is used.  (Previous usage is [here](https://github.com/apache/tvm/blob/main/src/relax/ir/expr.cc#L613), which defines the use of `CallNode::sinfo_args` as the default inferred struct info for external functions.)